{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "croissant_metadata = {\n",
    "  \"@context\": {\n",
    "    \"@vocab\": \"https://schema.org/\",\n",
    "    \"cr\": \"http://mlcommons.org/schema/\",\n",
    "    \"sc\": \"https://schema.org/\"\n",
    "  },\n",
    "  \"@type\": \"sc:Dataset\",\n",
    "  \"name\": \"PerturBench\",\n",
    "  \"description\": (\n",
    "    \"A dataset containing single-cell RNA-seq data with genetic and chemical \"\n",
    "    \"perturbations.\"\n",
    "  ),\n",
    "  \"url\": \"https://huggingface.co/datasets/altoslabs/perturbench\",\n",
    "  \"license\": \"https://creativecommons.org/licenses/by-nc/4.0/deed.en\",\n",
    "  \"citation\": (\n",
    "    \"Yan Wu, Esther Wershof, Sebastian M Schmon, Marcel Nassar, Błażej Osiński, \"\n",
    "    \"Ridvan Eksi, Zichao Yan, Rory Stark, Kun Zhang, and Thore Graepel (2025). \"\n",
    "    \"PerturBench: Benchmarking Machine Learning Models for Cellular Perturbation Analysis. \"\n",
    "    \"Proceedings of the 39th Conference on Neural Information Processing Systems (NeurIPS 2025).\"\n",
    "  ),\n",
    "  \"datePublished\": \"2025-05-15\",\n",
    "  \"version\": \"1.0\",\n",
    "  \"conformsTo\": \"http://mlcommons.org/croissant/1.0\",\n",
    "  \"distribution\": [\n",
    "    # File Objects (the actual files)\n",
    "    {\n",
    "      \"@type\": \"sc:FileObject\",\n",
    "      \"@id\": \"srivatsan20-h5ad\",\n",
    "      \"name\": \"srivatsan20_h5ad_file\",\n",
    "      \"description\": \"Gzipped HDF5 file for the Srivatsan20 dataset.\",\n",
    "      \"contentUrl\": \"https://huggingface.co/datasets/altoslabs/perturbench/resolve/main/srivatsan20_highest_dose_processed.h5ad.gz\",\n",
    "      \"encodingFormat\": \"application/gzip\",\n",
    "      \"sha256\": \"ddd46251ea3942c0e3799f64cada9718b455ffdeabb1a8fe8b3333ebec4946f0\"\n",
    "    },\n",
    "    {\n",
    "      \"@type\": \"sc:FileObject\",\n",
    "      \"@id\": \"norman19-h5ad\",\n",
    "      \"name\": \"norman19_h5ad_file\",\n",
    "      \"description\": \"Gzipped HDF5 file for the Norman19 dataset.\",\n",
    "      \"contentUrl\": \"https://huggingface.co/datasets/altoslabs/perturbench/resolve/main/norman19_processed.h5ad.gz\",\n",
    "      \"encodingFormat\": \"application/gzip\",\n",
    "      \"sha256\": \"8f880e3d85bb68a73a6d044cae13f62a86b8132b0302c6e1bc4b4e5564e6530f\"\n",
    "    },\n",
    "    {\n",
    "      \"@type\": \"sc:FileObject\",\n",
    "      \"@id\": \"norman19-cpa-h5ad\",\n",
    "      \"name\": \"norman19_cpa_h5ad_file\",\n",
    "      \"description\": \"CPA publication version of the norman19 dataset, subset to highly variable genes only.\",\n",
    "      \"contentUrl\": \"https://huggingface.co/datasets/altoslabs/perturbench/resolve/main/norman19_cpa_hvg_normalized_curated.h5ad.gz\",\n",
    "      \"encodingFormat\": \"application/gzip\",\n",
    "      \"sha256\": \"f04b53e1e2390d4e8b59029fc6e482f160a77b73c6965d6b487975954b76ad49\"\n",
    "    },\n",
    "    {\n",
    "      \"@type\": \"sc:FileObject\",\n",
    "      \"@id\": \"norman19-cpa-splits\",\n",
    "      \"name\": \"norman19_cpa_splits_file\",\n",
    "      \"description\": \"CPA publication splits for their version of the norman19 dataset\",\n",
    "      \"contentUrl\": \"https://huggingface.co/datasets/altoslabs/perturbench/resolve/main/norman19_cpa_hvg_normalized_splits.tar.gz\",\n",
    "      \"encodingFormat\": \"application/gzip\",\n",
    "      \"sha256\": \"d4369b3ca4864e6348ca8a43f77feae20b36fede7f3d9dfa8e6e31abd1156962\"\n",
    "    },\n",
    "    {\n",
    "      \"@type\": \"sc:FileObject\",\n",
    "      \"@id\": \"frangieh21-h5ad\",\n",
    "      \"name\": \"frangieh21_h5ad_file\",\n",
    "      \"description\": \"The gzipped HDF5 file containing the processed perturbation data.\",\n",
    "      \"contentUrl\": \"https://huggingface.co/datasets/altoslabs/perturbench/resolve/main/frangieh21_processed.h5ad.gz\",\n",
    "      \"encodingFormat\": \"application/gzip\",\n",
    "      \"sha256\": \"d8e93131e33c02b9fce9d1a1ba8e4924ae6cc4907341b45138ef82cb4a162c79\"\n",
    "    },\n",
    "    {\n",
    "      \"@type\": \"sc:FileObject\",\n",
    "      \"@id\": \"frangieh21-csv\",\n",
    "      \"name\": \"frangieh21_csv_file\",\n",
    "      \"description\": \"CSV file containing the data splits for the dataset.\",\n",
    "      \"contentUrl\": \"https://huggingface.co/datasets/altoslabs/perturbench/resolve/main/frangieh21_split.csv\",\n",
    "      \"encodingFormat\": \"text/csv\",\n",
    "      \"sha256\": \"ac2e8361a44239eb07aa813ca9025985b399aecce4851ef879a534bb1a72b711\"\n",
    "    },\n",
    "    {\n",
    "      \"@type\": \"sc:FileObject\",\n",
    "      \"@id\": \"mcfalinefigueroa23-h5ad\",\n",
    "      \"name\": \"mcfalinefigueroa23_h5ad_file\",\n",
    "      \"description\": \"Gzipped HDF5 file for the McFalineFigueroa23 dataset.\",\n",
    "      \"contentUrl\": \"https://huggingface.co/datasets/altoslabs/perturbench/resolve/main/mcfaline23_gxe_processed.h5ad.gz\",\n",
    "      \"encodingFormat\": \"application/gzip\",\n",
    "      \"sha256\": \"f6639e1ee8f24d2f2d0eea0011446fad370cfba96009ffc257b85fa604aa36f8\"\n",
    "    },\n",
    "    {\n",
    "      \"@type\": \"sc:FileObject\",\n",
    "      \"@id\": \"mcfalinefigueroa23-splits\",\n",
    "      \"name\": \"mcfalinefigueroa23_splits_file\",\n",
    "      \"description\": \"Gzipped tar archive containing the data splits for the McFalineFigueroa23 dataset. Each split corresponds to a different data scale (small, medium, full).\",\n",
    "      \"contentUrl\": \"https://huggingface.co/datasets/altoslabs/perturbench/resolve/main/mcfaline23_gxe_splits.tar.gz\",\n",
    "      \"encodingFormat\": \"application/gzip\",\n",
    "      \"sha256\": \"a383a9e8e38e63756e5dd9e1a4df7844a97669ef0d553c70de8f1441e40f0aa7\"\n",
    "    },\n",
    "    {\n",
    "      \"@type\": \"sc:FileObject\",\n",
    "      \"@id\": \"jiang24-h5ad\",\n",
    "      \"name\": \"jiang24_h5ad_file\",\n",
    "      \"description\": \"Gzipped HDF5 file for the Jiang24 dataset.\",\n",
    "      \"contentUrl\": \"https://huggingface.co/datasets/altoslabs/perturbench/resolve/main/jiang24_processed.h5ad.gz\",\n",
    "      \"encodingFormat\": \"application/gzip\",\n",
    "      \"sha256\": \"e92dec6610e261f31eb15d50aff0c12e19e99d992452e1f14d560477e741ccfe\"\n",
    "    },\n",
    "    {\n",
    "      \"@type\": \"sc:FileObject\",\n",
    "      \"@id\": \"jiang24-csv\",\n",
    "      \"name\": \"jiang24_csv_file\",\n",
    "      \"description\": \"CSV file containing the data split for the Jiang24 dataset.\",\n",
    "      \"contentUrl\": \"https://huggingface.co/datasets/altoslabs/perturbench/resolve/main/jiang24_split.csv\",\n",
    "      \"encodingFormat\": \"text/csv\",\n",
    "      \"sha256\": \"f1822ab1252b363727f30bbe8889384c2b306369889364390ecccb80e75f87d7\"\n",
    "    },\n",
    "    {\n",
    "      \"@type\": \"sc:FileObject\",\n",
    "      \"@id\": \"op3-h5ad\",\n",
    "      \"name\": \"op3_h5ad_file\",\n",
    "      \"description\": \"Gzipped HDF5 file for the OP3 dataset.\",\n",
    "      \"contentUrl\": \"https://huggingface.co/datasets/altoslabs/perturbench/resolve/main/op3_processed.h5ad.gz\",\n",
    "      \"encodingFormat\": \"application/gzip\",\n",
    "      \"sha256\": \"a95d7b3717128f89d34abdc5031e38bd04ed616296b9344f720f9e1e9d265b99\"\n",
    "    },\n",
    "    {\n",
    "      \"@type\": \"sc:FileObject\",\n",
    "      \"@id\": \"op3-csv\",\n",
    "      \"name\": \"op3_csv_file\",\n",
    "      \"description\": \"CSV file containing the data split for the OP3 dataset.\",\n",
    "      \"contentUrl\": \"https://huggingface.co/datasets/altoslabs/perturbench/resolve/main/op3_split.csv\",\n",
    "      \"encodingFormat\": \"text/csv\",\n",
    "      \"sha256\": \"433e23e944ba7865185520a41510323186f755f7bc5a5a413d11704e2db60e90\"\n",
    "    },\n",
    "    # File Sets (groups of files)\n",
    "    {\n",
    "      \"@type\": \"sc:FileSet\",\n",
    "      \"@id\": \"srivatsan20\",\n",
    "      \"name\": \"Srivatsan20\",\n",
    "      \"description\": (\n",
    "        \"This dataset is a modified version of the Srivatsan20 dataset published by \"\n",
    "        \"Srivatsan et al. via GEO:GSE139944. It contains 188 chemical perturbations \"\n",
    "        \"subset to the highest dose only applied across 3 cell types. The full data \"\n",
    "        \"preprocessing notebook can be found at: \"\n",
    "        \"https://github.com/altoslabs/perturbench/blob/main/notebooks/neurips2025/\"\n",
    "        \"data_curation/curate_Srivatsan20.ipynb.\"\n",
    "      ),\n",
    "      \"encodingFormat\": \"application/gzip\",\n",
    "      \"cr:includes\": [\n",
    "        \"srivatsan20-h5ad\"\n",
    "      ]\n",
    "    },\n",
    "    {\n",
    "      \"@type\": \"sc:FileSet\",\n",
    "      \"@id\": \"norman19\",\n",
    "      \"name\": \"Norman19\",\n",
    "      \"description\": (\n",
    "        \"This dataset is a modified version of the Norman19 dataset published by \"\n",
    "        \"Norman et al. via GEO:GSE133344. It contains 287 genetic perturbations \"\n",
    "        \"(131 duals) applied to k562 cells. The full data preprocessing notebook can \"\n",
    "        \"be found at: https://github.com/altoslabs/perturbench/blob/main/notebooks/\"\n",
    "        \"neurips2025/data_curation/curate_Norman19.ipynb.\"\n",
    "      ),\n",
    "      \"encodingFormat\": \"application/gzip\",\n",
    "      \"cr:includes\": [\n",
    "        \"norman19-h5ad\",\n",
    "        \"norman19-cpa-h5ad\",\n",
    "        \"norman19-cpa-splits\"\n",
    "      ]\n",
    "    },\n",
    "    {\n",
    "      \"@type\": \"sc:FileSet\",\n",
    "      \"@id\": \"frangieh21\",\n",
    "      \"name\": \"Frangieh21\",\n",
    "      \"description\": (\n",
    "        \"This dataset is a modified version of the Frangieh21 dataset published by \"\n",
    "        \"Frangieh et al. via https://singlecell.broadinstitute.org/single_cell/\"\n",
    "        \"study/SCP1064/multi-modal-pooled-perturb-cite-seq-screens-in-patient-models-\"\n",
    "        \"define-novel-mechanisms-of-cancer-immune-evasion. It contains 248 genetic \"\n",
    "        \"perturbations applied to 3 melanoma cell models. The full data preprocessing \"\n",
    "        \"notebook can be found at: https://github.com/altoslabs/perturbench/blob/main/\"\n",
    "        \"notebooks/neurips2025/data_curation/curate_Frangieh21.ipynb.\"\n",
    "      ),\n",
    "      \"encodingFormat\": \"application/gzip\",\n",
    "      \"cr:includes\": [\n",
    "        \"frangieh21-h5ad\",\n",
    "        \"frangieh21-csv\"\n",
    "      ]\n",
    "    },\n",
    "    {\n",
    "      \"@type\": \"sc:FileSet\",\n",
    "      \"@id\": \"mcfalinefigueroa23\",\n",
    "      \"name\": \"McFalineFigueroa23\",\n",
    "      \"description\": (\n",
    "        \"This dataset is a modified version of the McFalineFigueroa23 dataset \"\n",
    "        \"published by McFaline-Figueroa et al. via GEO:GSE225775. It contains ~200 \"\n",
    "        \"perturbations applied across 6 cell lines and 5 cytokine treatments (30 \"\n",
    "        \"unique biological states). The data preprocessing occured in two steps and \"\n",
    "        \"both files can be found at: https://github.com/altoslabs/perturbench/blob/\"\n",
    "        \"main/notebooks/neurips2025/data_curation/ with the \"\n",
    "        \"curate_McFalineFigueroa_2023 prefix.\"\n",
    "      ),\n",
    "      \"encodingFormat\": \"application/gzip\",\n",
    "      \"cr:includes\": [\n",
    "        \"mcfalinefigueroa23-h5ad\",\n",
    "        \"mcfalinefigueroa23-splits\"\n",
    "      ]\n",
    "    },\n",
    "    {\n",
    "      \"@type\": \"sc:FileSet\",\n",
    "      \"@id\": \"jiang24\",\n",
    "      \"name\": \"Jiang24\",\n",
    "      \"description\": (\n",
    "        \"This dataset is a modified version of the Jiang24 dataset published by \"\n",
    "        \"Jiang et al. via https://zenodo.org/records/14518762. It contains 525 \"\n",
    "        \"genetic perturbations applied across 3 cell lines and 5 chemical treatments \"\n",
    "        \"(15 unique biological states). The data preprocessing occured in two steps \"\n",
    "        \"and both files can be found at: https://github.com/altoslabs/perturbench/\"\n",
    "        \"blob/main/notebooks/neurips2025/data_curation/ with the curate_Jiang_2024 \"\n",
    "        \"prefix.\"\n",
    "      ),\n",
    "      \"encodingFormat\": \"application/gzip\",\n",
    "      \"cr:includes\": [\n",
    "        \"jiang24-h5ad\",\n",
    "        \"jiang24-csv\"\n",
    "      ]\n",
    "    },\n",
    "    {\n",
    "      \"@type\": \"sc:FileSet\",\n",
    "      \"@id\": \"op3\",\n",
    "      \"name\": \"OP3\",\n",
    "      \"description\": (\n",
    "        \"This dataset is a modified version of the OpenProblems perturbation \"\n",
    "        \"prediction challenge dataset that a Kaggle competition part of the NeurIPS \"\n",
    "        \"2023 competition track by Burkhardt et al. via \"\n",
    "        \"https://openproblems.bio/benchmarks/perturbation_prediction?version=v1.0.0. \"\n",
    "        \"It contains 144 chemical perturbations applied to PBMCs with at least 4 \" # Corrected here\n",
    "        \"mature cell types. The data preprocessing can be found at: \"\n",
    "        \"https://github.com/altoslabs/perturbench/blob/main/notebooks/neurips2025/data_curation/curate_op3.ipynb.\"\n",
    "      ),\n",
    "      \"encodingFormat\": \"application/gzip\",\n",
    "      \"cr:includes\": [\n",
    "        \"op3-h5ad\",\n",
    "        \"op3-csv\"\n",
    "      ]\n",
    "    }\n",
    "  ],\n",
    "  \"recordSet\": [\n",
    "    {\n",
    "      \"@type\": \"cr:RecordSet\",\n",
    "      \"name\": \"treatment_classification\",\n",
    "      \"description\": (\n",
    "        \"Records for the perturbation response prediction task using scRNA-seq \"\n",
    "        \"datasets stored in h5ad files. Records are sourced from six manifest files \" # Corrected here\n",
    "        \"found in the distribution list, with one file per dataset, and stored in \"\n",
    "        \"the `.obs` slot of each h5ad file. The perturbation identity is the \"\n",
    "        \"'condition' column, with the `control` value reserved for the DMSO or \"\n",
    "        \"non-targeting CRISPR controls. The cell type is stored in the `cell_type` \"\n",
    "        \"column. Additional cytokine or chemical treatments are stored in the \"\n",
    "        \"`treatment` column. Splits are defined dynamically by the PerturBench \"\n",
    "        \"library.\"\n",
    "      ),\n",
    "      \"field\": [\n",
    "        {\n",
    "          \"@type\": \"cr:Field\",\n",
    "          \"name\": \"condition\",\n",
    "          \"description\": (\n",
    "            \"Chemical or genetic perturbation applied to cells. This is the key \"\n",
    "            \"conditioning label for the perturbation response prediction task.\"\n",
    "          ),\n",
    "          \"dataType\": \"sc:Text\"\n",
    "        },\n",
    "        {\n",
    "          \"@type\": \"cr:Field\",\n",
    "          \"name\": \"cell_type\",\n",
    "          \"description\": (\n",
    "            \"Cell line or cell type of the sample.\"\n",
    "          ),\n",
    "          \"dataType\": \"sc:Text\"\n",
    "        },\n",
    "        {\n",
    "          \"@type\": \"cr:Field\",\n",
    "          \"name\": \"treatment\",\n",
    "          \"description\": (\n",
    "            \"Additional cytokine or chemical treatments applied to cells, used \"\n",
    "            \"together with `cell_type` to define the biological state of a sample.\"\n",
    "          ),\n",
    "          \"dataType\": \"sc:Text\"\n",
    "        }\n",
    "      ]\n",
    "    }\n",
    "  ]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Croissant metadata saved to croissant.json\n"
     ]
    }
   ],
   "source": [
    "# Save the Croissant metadata to a JSON file\n",
    "output_file = \"croissant.json\"\n",
    "\n",
    "# Write the JSON to the output file\n",
    "with open(output_file, 'w') as f:\n",
    "    json.dump(croissant_metadata, f, indent=2)\n",
    "\n",
    "print(f\"Croissant metadata saved to {output_file}\")"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 2
}
